File: test_perspective.py
import math

print("Starting Perspective Unit Tests (Python)...")

# Mock constants
COLUMN_COUNT = 16
COLUMN_WIDTH = 320
GAP = 40
RADIUS = 1800

# Test 1: Angle Step Calculation
# JS: (2 * Math.atan((COLUMN_WIDTH + GAP) / (2 * RADIUS))) * (180 / Math.PI)
angle_step_rad = 2 * math.atan((COLUMN_WIDTH + GAP) / (2 * RADIUS))
angle_step_deg = angle_step_rad * (180 / math.pi)

print(f"Test 1: Angle Step")
print(f"- Expected: ~11.4 degrees")
print(f"- Actual: {angle_step_deg:.4f} degrees")

if 10 < angle_step_deg < 13:
    print("PASS: Angle step within reasonable range.")
else:
    print("FAIL: Angle step calculation looks wrong.")

# Test 2: Max Angle (Bounds)
TOTAL_ARC = COLUMN_COUNT * angle_step_deg
VISIBLE_ARC = 60
MAX_ANGLE = max(0, (TOTAL_ARC - VISIBLE_ARC) / 2)

print(f"\nTest 2: Max Angle (Bounds)")
print(f"- Total Arc: {TOTAL_ARC:.1f} degrees")
print(f"- Max Angle: {MAX_ANGLE:.1f} degrees")

if 0 < MAX_ANGLE < 180:
    print("PASS: Max angle is positive and valid.")
else:
    print("FAIL: Max angle invalid.")

# Test 3: Infinite Scroll Logic
SET_HEIGHT = 1000
current_scroll_y = 0
target_scroll_y = 0

print(f"\nTest 3: Infinite Scroll Logic")

# Simulate scroll down
current_scroll_y = SET_HEIGHT + 10
target_scroll_y = SET_HEIGHT + 10

if current_scroll_y > SET_HEIGHT:
    current_scroll_y -= SET_HEIGHT
    target_scroll_y -= SET_HEIGHT

print(f"- After Scroll Reset: current={current_scroll_y}, target={target_scroll_y}")
if current_scroll_y == 10 and target_scroll_y == 10:
    print("PASS: Scroll wrap-around (down) works.")
else:
    print("FAIL: Scroll wrap-around (down) failed.")

print("\nAll tests complete.")

File: test_perspective.js
// Unit Test for Perspective Calculations
// To run: node test_perspective.js

console.log("Starting Perspective Unit Tests...");

// Mock constants from grid-phantom.html
const COLUMN_COUNT = 16; 
const COLUMN_WIDTH = 320; 
const GAP = 40;
const RADIUS = 1800; 

// Test 1: Angle Step Calculation
// angleStep = (2 * Math.atan((COLUMN_WIDTH + GAP) / (2 * RADIUS))) * (180 / Math.PI);
const angleStepRad = 2 * Math.atan((COLUMN_WIDTH + GAP) / (2 * RADIUS));
const angleStepDeg = angleStepRad * (180 / Math.PI);

console.log(`Test 1: Angle Step`);
console.log(`- Expected: ~11.4 degrees`); // Rough calc: 360/1800 rad ~ 0.2 rad ~ 11 deg
console.log(`- Actual: ${angleStepDeg.toFixed(4)} degrees`);

if (angleStepDeg > 10 && angleStepDeg < 13) {
    console.log("PASS: Angle step within reasonable range.");
} else {
    console.error("FAIL: Angle step calculation looks wrong.");
}

// Test 2: Max Angle (Bounds)
const TOTAL_ARC = COLUMN_COUNT * angleStepDeg;
const VISIBLE_ARC = 60; 
const MAX_ANGLE = Math.max(0, (TOTAL_ARC - VISIBLE_ARC) / 2);

console.log(`\nTest 2: Max Angle (Bounds)`);
console.log(`- Total Arc: ${TOTAL_ARC.toFixed(1)} degrees`);
console.log(`- Max Angle: ${MAX_ANGLE.toFixed(1)} degrees`);

// If we have 16 columns * 11 deg = 176 deg total.
// Visible is 60. 
// Max rotation = (176 - 60)/2 = 58 deg.
if (MAX_ANGLE > 0 && MAX_ANGLE < 180) {
    console.log("PASS: Max angle is positive and valid.");
} else {
    console.error("FAIL: Max angle invalid.");
}

// Test 3: Infinite Scroll Logic
const SET_HEIGHT = 1000; // Mock height
let currentScrollY = 0;
let targetScrollY = 0;

console.log(`\nTest 3: Infinite Scroll Logic`);

// Simulate scroll down past set height
currentScrollY = SET_HEIGHT + 10;
targetScrollY = SET_HEIGHT + 10;

// Apply logic
if (currentScrollY > SET_HEIGHT) {
    currentScrollY -= SET_HEIGHT;
    targetScrollY -= SET_HEIGHT;
}

console.log(`- After Scroll Reset: current=${currentScrollY}, target=${targetScrollY}`);
if (currentScrollY === 10 && targetScrollY === 10) {
    console.log("PASS: Scroll wrap-around (down) works.");
} else {
    console.error("FAIL: Scroll wrap-around (down) failed.");
}

// Simulate scroll up past negative set height
currentScrollY = -SET_HEIGHT - 20;
targetScrollY = -SET_HEIGHT - 20;

if (currentScrollY < -SET_HEIGHT) {
    currentScrollY += SET_HEIGHT;
    targetScrollY += SET_HEIGHT;
}

console.log(`- After Scroll Reset (Up): current=${currentScrollY}, target=${targetScrollY}`);
if (currentScrollY === -20 && targetScrollY === -20) {
    console.log("PASS: Scroll wrap-around (up) works.");
} else {
    console.error("FAIL: Scroll wrap-around (up) failed.");
}

console.log("\nAll tests complete.");